cli/mender-artifact/debugfs.go

```diff
@@ -104,53 +105,100 @@ func debugfsReplaceFile(imageFile, newFile, image string) (err error) {
 	executeCommand(cmd, image)
 	// Write to the partition
 	cmd = fmt.Sprintf("cd %s\nwrite %s %s\nclose", filepath.Dir(imageFile), newFile, filepath.Base(imageFile))
-	return executeCommand(cmd, image)
+	_, err = executeCommand(cmd, image)
+	return err
+}
+
+func debugfsRemoveFileOrDir(imageFile, image string, recursive bool) (err error) {
+	// Check that the file or directory exists.
+	cmd := fmt.Sprintf("cd %s", filepath.Dir(imageFile))
+	if _, err = executeCommand(cmd, image); err != nil {
+		return err
+	}
+	isDir := filepath.Dir(imageFile) == filepath.Clean(imageFile)
+	var deleteCmd string
+	if isDir {
+		return debugfsRemoveDir(imageFile, image, recursive)
+	} else {
+		deleteCmd = "rm"
+	}
+	cmd = fmt.Sprintf("%s %s\nclose", deleteCmd, imageFile)
+	_, err = executeCommand(cmd, image)
+	return err
+}
+
+func debugfsRemoveDir(imageFile, image string, recursive bool) (err error) {
+	// Remove the `/` suffix if present, as debugfs does not play nice with it.
+	imageFile = strings.TrimRight(imageFile, "/")
+	cmd := fmt.Sprintf("rmdir %s", imageFile)
+	if _, err = executeCommand(cmd, image); err != nil {
+		if recursive && strings.Contains(err.Error(), "directory not empty") {
+			// Recurse and remove all files in the directory.
+			cmd = fmt.Sprintf("ls %s", imageFile)
+			buf, err := executeCommand(cmd, image)
+			if err != nil {
+				return errors.Wrap(err, "debugfsRemoveDir")
+			}
+			regexp := regexp.MustCompile(`[0-9]+ +\([0-9]+\) +((.\w+.)+)`)
+			for _, filename := range regexp.FindAllStringSubmatch(buf.String(), -1) {
+				err = debugfsRemoveFileOrDir(filepath.Join(imageFile, string(filename[1])), image, recursive)
+				if err != nil {
+					return errors.Wrap(err, "debugfsRemoveDir")
+				}
+			}
+			// Now that all the subdirs and files should be removed, try and remove the directory
+			// once more.
+			return debugfsRemoveDir(imageFile, image, recursive)
```

Comment:
Shouldn't this last call be with recursive=false? As the comment says, in theory there is no more files or subdirs, so we shouldn't need to recursive here in any case.